src/provisioningapi.ts

```diff
@@ -0,0 +1,153 @@
+	}
+
+	get v1(): Router {
+		return this.apiRouterV1;
+	}
+
+	private async checkProvisioningSharedSecret(req: IAuthedRequest, res: Response, next: () => void) {
+		if (!this.apiSharedSecret) {
+			res.status(FORBIDDEN).json({
+				error: "The provisioning API is disabled",
+				errcode: "M_FORBIDDEN",
+			});
+		} else if (req.header("Authorization") !== `Bearer ${this.apiSharedSecret}`) {
+			res.status(UNAUTHORIZED).json({
+				error: "Unknown or missing token",
+				errcode: "M_UNKNOWN_TOKEN",
+			});
+		} else if (!req.query.user_id) {
+			res.status(BAD_REQUEST).json({
+				error: "Missing user_id query parameter",
+				errcode: "M_BAD_REQUEST",
+			});
+		} else {
+			req.userId = req.query.user_id;
+			next();
+		}
+	}
+
+	private async status(req: IAuthedRequest, res: Response) {
+		const puppets = await this.bridge.provisioner.getForMxid(req.userId) as IPuppetWithDescription[];
+		if (this.bridge.hooks.getDesc) {
+			for (const data of puppets) {
+				data.description = await this.bridge.hooks.getDesc(data.puppetId, data.data);
+			}
+		}
+		res.json({
+			puppets,
+			permissions: {
+				create: this.bridge.provisioner.canCreate(req.userId),
+				relay: this.bridge.provisioner.canRelay(req.userId),
+			},
+		});
+	}
+
+	private async link(req: IAuthedRequest, res: Response) {
+		const puppetId = await this.bridge.provisioner.new(req.userId, req.body.data, req.body.remote_user_id);
+		res.status(CREATED).json({ puppet_id: puppetId });
+	}
+
+	private async getPuppetId(req: IAuthedRequest, res: Response): Promise<number | null> {
+		const puppetId = Number(req.params.puppetId);
+		const data = await this.bridge.provisioner.get(puppetId);
+		if (!data || data.puppetMxid !== req.userId) {
+			res.status(FORBIDDEN).json({
+				error: "You must own the puppet ID",
+				errcode: "M_FORBIDDEN",
+			});
+			return null;
+		}
+		return puppetId;
+	}
+
+	private async unlink(req: IAuthedRequest, res: Response) {
+		const puppetId = await this.getPuppetId(req, res);
+		if (!puppetId) {
+			return;
+		}
+		await this.bridge.provisioner.delete(req.userId, puppetId);
+		res.status(NO_CONTENT);
+	}
+
+	private async listUsers(req: IAuthedRequest, res: Response) {
+		if (!this.bridge.hooks.listUsers) {
+			res.status(NOT_IMPLEMENTED).json({
+				error: "listUsers hook not implemented",
+				errcode: "M_NOT_IMPLEMENTED",
+			});
+			return;
+		}
+		const puppetId = await this.getPuppetId(req, res);
+		if (!puppetId) {
+			return;
+		}
+		res.status(OK).json(await this.bridge.hooks.listUsers(puppetId));
+	}
+
+	private async listRooms(req: IAuthedRequest, res: Response) {
+		if (!this.bridge.hooks.listRooms) {
+			res.status(NOT_IMPLEMENTED).json({
+				error: "listUsers hook not implemented",
+				errcode: "M_NOT_IMPLEMENTED",
+			});
+			return;
+		}
+		const puppetId = await this.getPuppetId(req, res);
+		if (!puppetId) {
+			return;
+		}
+		res.status(OK).json(await this.bridge.hooks.listRooms(+req.params.puppetId));
```

Comment:
`Number(req.params.puppetId)`? Might be a bit more to type but it is immediately clear what it does

Reply:
Ohwait, you forgot to use the `puppetId` variable? ^^

Reply:
Oops, it's supposed to use `puppetId`